internal/server/instance/instance_utils.go

```diff
@@ -1257,3 +1258,97 @@ func SnapshotProtobufToInstanceArgs(s *state.State, inst Instance, snap *migrati
 
 	return &args, nil
 }
+
+// AllowedInstanceDevices checks whether custom block volumes are allowed to be attached to the given instance.
+// This is done by checking whether the provided devices are already attached to other instances.
+func AllowedInstanceDevices(s *state.State, instanceID int, projectName string, devices ...deviceConfig.Devices) error {
+	volumeDevices := make(map[int64][]cluster.Device)
+
+	err := s.DB.Cluster.Transaction(s.ShutdownCtx, func(ctx context.Context, tx *db.ClusterTx) error {
```

Comment:
This seems extremely inefficient and will significantly slow down startup in busy environments.

I suspect we'd get much better performance by moving this check from instance startup over to instance config validation.

Basically:
 - Refuse adding a block custom volume to a profile unless security.shared=true is set
 - Refuse adding a block custom volume to an instance if it's already present on another instance
 - Refuse setting security.shared=false on a custom volume that's currently used by more than one instance

Note that your check was also incorrect as you need to keep in mind that volume `foo` in pool `bar` on server `s1` is different from volume `foo` in pool `bar` on server `s2`.

With those checks being at configuration time rather than start time, the performance impact should be significantly lower and we'll save a lot of DB queries on the hot path.